chain/chain/src/runtime/mod.rs

```diff
@@ -777,23 +777,38 @@ impl RuntimeAdapter for NightshadeRuntime {
             usize::MAX
         };
 
-        // In general, we limit the number of transactions via send_fees.
-        // However, as a second line of defense, we want to limit the byte size
-        // of transaction as well. Rather than introducing a separate config for
-        // the limit, we compute it heuristically from the gas limit and the
-        // cost of roundtripping a byte of data through disk. For today's value
-        // of parameters, this corresponds to about 13megs worth of
-        // transactions.
-        let size_limit = transactions_gas_limit
-            / (runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::storage_write_value_byte)
-                + runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::storage_read_value_byte));
+        let size_limit: u64 =
+            if checked_feature!("stable", WitnessTransactionLimits, protocol_version) {
+                // Sum of transactions in the previous and current chunks should not exceed the limit.
+                // Witness keeps transactions from both previous and current chunk, so we have to limit the sum of both.
+                runtime_config
+                    .max_transactions_size_in_witness
+                    .saturating_sub(chunk.prev_chunk_transactions_size) as u64
+            } else {
+                // In general, we limit the number of transactions via send_fees.
+                // However, as a second line of defense, we want to limit the byte size
+                // of transaction as well. Rather than introducing a separate config for
+                // the limit, we compute it heuristically from the gas limit and the
+                // cost of roundtripping a byte of data through disk. For today's value
+                // of parameters, this corresponds to about 13megs worth of
+                // transactions.
+                transactions_gas_limit
+                    / (runtime_config
+                        .wasm_config
+                        .ext_costs
+                        .gas_cost(ExtCosts::storage_write_value_byte)
+                        + runtime_config
+                            .wasm_config
+                            .ext_costs
+                            .gas_cost(ExtCosts::storage_read_value_byte))
```

Comment:
nit: Can you split this into multiple lines by moving the factors / addends to helper variables? It's getting a bit too crazy ;)